DEVELOPERS_GUIDE.md

```diff
@@ -0,0 +1,203 @@
+- Include statements must be located at the top of a file only; except for template class implementations.
+
+### Statements
+
+- The parts of a class must be sorted public, protected and private.
+- Within the parts, methods should go before variables.
+- Use the ``explicit`` keyword for single argument constructors.
+- Type conversions must always be done explicitly, never rely on implicit type conversion. The only exception is upcasting, which may be done implicitly or explicitly.
+- For every ``new``, there must be a ``delete`` in the same class. If the ``delete`` is in another class, it must be documented carefully with both the ``new`` and ``delete``.
+- Variables must never have dual meaning.
+- Multiple variables should not be declared on the same line; except for the primitive data types e.g. ``int``.
+- Use of global variables should be minimized.
+- Class variables should never be declared public.
+- C++ pointers and references should have their reference symbol next to the name rather than to the type, i.e. ``Element *element``.
+- Implicit test for 0 should not be used other than for boolean variables.
+- Variables should be declared in the smallest scope possible.
+- Only loop control statements must be included in the ``for()`` construction.
+- Range based for-loops are preferred over other types of for-loops. For example, a for-loop over vector vec can look like ``for ( double d: vec) { }``.
+- Loop variables should be initialized immediately before the loop.
+- Do-while loops can be avoided.
+- The use of ``break`` and ``continue`` in loops should be avoided, go for readability.
+- The form ``while (true)`` should be used for infinite loops.
+- In cases of an if-statement, the nominal case should be put in the if-part and the exception in the else-part.
+- Floating point constants should always be written with decimal point and at least one decimal. (i.e. ``x = 0.0``).
+- Use ``nullptr`` instead of ``NULL`` or ``0``.
+- Try to avoid smart pointers. If a smart pointer is necessary, use unique_ptr or shared_ptr.
+- Use the type ``std::size_t`` for non-negative integer variables.
+- Encapsulate global variables and constants, enumerated types, and type definitions in a class or namespace.
+- The ``auto``-keyword for type declaration is only allowed if it is instantly clear what the type is.
+- When using a template, use typename T and NOT class T.
+- Avoid using ``typename`` as a return specifier.
+- Try to avoid ``typedef`` for type definitions. Use the ``using``-keyword if either the type is likely to change, or the typename is very long. This should be done in the class declaration and not globally.
+- Avoid returning by final function argument (for example ``void myFun (int a, int b, Matrix mat)`` for computing mat).
+
+
+
+## About assertions
+
+The intended use of assertions in general and ``logger.assert`` for hpGEM is to check blatant assumptions on the arguments provided to the
```

Comment:
We have changed `logger.assert` and now have two options, `logger.assert_debug` and `logger.assert_always`. The first one is compiled out in release mode. So I think we might need to rewrite this section a bit. My general feel is that
 
 - assert_always is where removing the check is not expected to make any (significant) speed gain or things that must always be checked. Thus, usage is restricted to checks that are cheap in code that runs infrequently. Furthermore, it is used in unittests.
 - assert_debug, used everywhere else in liberal fashion.